// ngx-echarts/bundles/ngx-echarts.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs/Observable'), require('rxjs/add/observable/of'), require('rxjs/add/observable/empty')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', 'rxjs/Observable', 'rxjs/add/observable/of', 'rxjs/add/observable/empty'], factory) :
	(factory((global['ngx-echarts'] = {}),global.ng.core,global.Rx));
}(this, (function (exports,core,Observable) { 'use strict';

var ChangeFilter = (function () {
    /**
     * @param {?} _changes
     */
    function ChangeFilter(_changes) {
        this._changes = _changes;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ChangeFilter.of = function (changes) {
        return new ChangeFilter(changes);
    };
    /**
     * @template T
     * @param {?} key
     * @return {?}
     */
    ChangeFilter.prototype.notEmpty = function (key) {
        if (this._changes[key]) {
            var /** @type {?} */ value = this._changes[key].currentValue;
            if (value !== undefined && value !== null) {
                return Observable.Observable.of(value);
            }
        }
        return Observable.Observable.empty();
    };
    /**
     * @template T
     * @param {?} key
     * @return {?}
     */
    ChangeFilter.prototype.has = function (key) {
        if (this._changes[key]) {
            var /** @type {?} */ value = this._changes[key].currentValue;
            return Observable.Observable.of(value);
        }
        return Observable.Observable.empty();
    };
    return ChangeFilter;
}());
var NgxEchartsDirective = (function () {
    /**
     * @param {?} el
     * @param {?} _ngZone
     */
    function NgxEchartsDirective(el, _ngZone) {
        this.el = el;
        this._ngZone = _ngZone;
        // chart events:
        this.chartInit = new core.EventEmitter();
        this.chartClick = new core.EventEmitter();
        this.chartDblClick = new core.EventEmitter();
        this.chartMouseDown = new core.EventEmitter();
        this.chartMouseUp = new core.EventEmitter();
        this.chartMouseOver = new core.EventEmitter();
        this.chartMouseOut = new core.EventEmitter();
        this.chartGlobalOut = new core.EventEmitter();
        this.chartContextMenu = new core.EventEmitter();
        this.chartDataZoom = new core.EventEmitter();
        this._chart = null;
        this.currentWindowWidth = null;
    }
    /**
     * @return {?}
     */
    NgxEchartsDirective.prototype.createChart = function () {
        var _this = this;
        this.currentWindowWidth = window.innerWidth;
        var /** @type {?} */ dom = this.el.nativeElement;
        if (window && window.getComputedStyle) {
            var /** @type {?} */ prop = window.getComputedStyle(dom, null).getPropertyValue('height');
            if (!prop || prop === '0px') {
                dom.style.height = '400px';
            }
        }
        return this._ngZone.runOutsideAngular(function () { return echarts.init(dom, _this.theme || undefined, _this.initOpts || undefined); });
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NgxEchartsDirective.prototype.onWindowResize = function (event) {
        if (event.target.innerWidth !== this.currentWindowWidth) {
            this.currentWindowWidth = event.target.innerWidth;
            if (this._chart) {
                this._chart.resize();
            }
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    NgxEchartsDirective.prototype.ngOnChanges = function (changes) {
        var _this = this;
        var /** @type {?} */ filter = ChangeFilter.of(changes);
        filter.notEmpty('options').subscribe(function (opt) { return _this.onOptionsChange(opt); });
        filter.notEmpty('merge').subscribe(function (opt) { return _this.setOption(opt); });
        filter.has('loading').subscribe(function (v) { return _this.toggleLoading(!!v); });
    };
    /**
     * @return {?}
     */
    NgxEchartsDirective.prototype.ngOnDestroy = function () {
        if (this._chart) {
            this._chart.dispose();
            this._chart = null;
        }
    };
    /**
     * @param {?} opt
     * @return {?}
     */
    NgxEchartsDirective.prototype.onOptionsChange = function (opt) {
        if (opt) {
            if (!this._chart) {
                this._chart = this.createChart();
                // output echart instance:
                this.chartInit.emit(this._chart);
                // register events:
                this.registerEvents(this._chart);
            }
            this._chart.setOption(this.options, true);
            this._chart.resize();
        }
    };
    /**
     * @param {?} _chart
     * @return {?}
     */
    NgxEchartsDirective.prototype.registerEvents = function (_chart) {
        var _this = this;
        if (_chart) {
            // register mouse events:
            _chart.on('click', function (e) { return _this.chartClick.emit(e); });
            _chart.on('dblClick', function (e) { return _this.chartDblClick.emit(e); });
            _chart.on('mousedown', function (e) { return _this.chartMouseDown.emit(e); });
            _chart.on('mouseup', function (e) { return _this.chartMouseUp.emit(e); });
            _chart.on('mouseover', function (e) { return _this.chartMouseOver.emit(e); });
            _chart.on('mouseout', function (e) { return _this.chartMouseOut.emit(e); });
            _chart.on('globalout', function (e) { return _this.chartGlobalOut.emit(e); });
            _chart.on('contextmenu', function (e) { return _this.chartContextMenu.emit(e); });
            // other events;
            _chart.on('datazoom', function (e) { return _this.chartDataZoom.emit(e); });
        }
    };
    /**
     * @return {?}
     */
    NgxEchartsDirective.prototype.clear = function () {
        if (this._chart) {
            this._chart.clear();
        }
    };
    /**
     * @param {?} loading
     * @return {?}
     */
    NgxEchartsDirective.prototype.toggleLoading = function (loading) {
        if (this._chart) {
            loading ? this._chart.showLoading() : this._chart.hideLoading();
        }
    };
    /**
     * @param {?} option
     * @param {?=} opts
     * @return {?}
     */
    NgxEchartsDirective.prototype.setOption = function (option, opts) {
        if (this._chart) {
            this._chart.setOption(option, opts);
        }
    };
    return NgxEchartsDirective;
}());
NgxEchartsDirective.decorators = [
    { type: core.Directive, args: [{
                selector: 'echarts, [echarts]'
            },] },
];
/**
 * @nocollapse
 */
NgxEchartsDirective.ctorParameters = function () { return [
    { type: core.ElementRef, },
    { type: core.NgZone, },
]; };
NgxEchartsDirective.propDecorators = {
    'options': [{ type: core.Input },],
    'theme': [{ type: core.Input },],
    'loading': [{ type: core.Input },],
    'initOpts': [{ type: core.Input },],
    'merge': [{ type: core.Input },],
    'chartInit': [{ type: core.Output },],
    'chartClick': [{ type: core.Output },],
    'chartDblClick': [{ type: core.Output },],
    'chartMouseDown': [{ type: core.Output },],
    'chartMouseUp': [{ type: core.Output },],
    'chartMouseOver': [{ type: core.Output },],
    'chartMouseOut': [{ type: core.Output },],
    'chartGlobalOut': [{ type: core.Output },],
    'chartContextMenu': [{ type: core.Output },],
    'chartDataZoom': [{ type: core.Output },],
    'onWindowResize': [{ type: core.HostListener, args: ['window:resize', ['$event'],] },],
};
var NgxEchartsModule = (function () {
    function NgxEchartsModule() {
    }
    return NgxEchartsModule;
}());
NgxEchartsModule.decorators = [
    { type: core.NgModule, args: [{
                declarations: [
                    NgxEchartsDirective
                ],
                exports: [
                    NgxEchartsDirective
                ]
            },] },
];
/**
 * @nocollapse
 */
NgxEchartsModule.ctorParameters = function () { return []; };

exports.NgxEchartsModule = NgxEchartsModule;
exports.NgxEchartsDirective = NgxEchartsDirective;

Object.defineProperty(exports, '__esModule', { value: true });

})));
